### PR TITLE
Clickable code search results line numbers.

### LINKED ISSUE #1580: A "SearchASM.csx" script returns wrong line numbers sometimes.
### Describe the bug

A "SearchASM.csx" script returns wrong line numbers for the code entries with blank lines in "Disassembly".

### Reproducing steps

1. Use the script with such search query that returns the code entries with blank lines in "Disassembly".

### Setup Details

1. UTMT version - https://github.com/krzys-h/UndertaleModTool/commit/34bebb63984fa54bf46c4798e6dd92373fc5cd32

### PR DESCRIPTION
**Pull request by VladiStep**
## Description
1) **Made the "Line (number)" part of the code search results lines clickable.**
2) Moved code editor related property (`CodeEditorDecompile`) and enum (`CodeEditorMode`) out from the `MainWindow`, and also renamed them.
3) Fixed some variable names in the code search scripts.
4) Added an optional `overwrite` argument for `Tab.SaveTabContentState()`.

image

Fixes #1580.

### CODE REVIEW

File: UndertaleModCli/Program.UMTLibInherited.cs [hunk 1]
```diff
@@ -521,13 +521,13 @@ public string SimpleTextInput(string title, string label, string defaultValue, b
     }
 
     /// <inheritdoc/>
-    public async Task ClickableSearchOutput(string title, string query, int resultsCount, IOrderedEnumerable<KeyValuePair<string, List<string>>> resultsDict, bool editorDecompile, IOrderedEnumerable<string> failedList = null)
+    public async Task ClickableSearchOutput(string title, string query, int resultsCount, IOrderedEnumerable<KeyValuePair<string, List<(int, string)>>> resultsDict, bool editorDecompile, IOrderedEnumerable<string> failedList = null)
```
Comment (Miepee) [inline UndertaleModCli/Program.UMTLibInherited.cs:524]: Seeing a tuple here feels very weird to me right now, but I currently can't think of something nicer. 
In the meantime, can you at least use names for them? `List<(int lineNumber, string codeSnippet)>` or something

File: UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs [hunk 2]
```diff
@@ -375,15 +385,45 @@ private static void RestoreCaretPosition(TextEditor textEditor, int linePos, int
...
         }
-        
+        public static void ChangeLineNumber(int lineNum, CodeEditorTab editorTab)
+        {
+            if (lineNum < 1)
+                return;
+
+            if (editorTab == CodeEditorTab.Unstated)
+            {
+                Debug.WriteLine("The \"editorTab\" argument of \"GoToLine()\" is \"Unstated\".");
```
Comment (Miepee) [inline UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs:404]: using `nameof()` is probably nicer here.

File: UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs [hunk 3]
```diff
@@ -375,15 +385,45 @@ private static void RestoreCaretPosition(TextEditor textEditor, int linePos, int
...
+                OverriddenDisasmPos = (lineNum, -1, -1);
+        }
+        public static void ChangeLineNumber(int lineNum, TextEditor textEditor)
+        {
+            if (lineNum < 1)
+                return;
+
+            if (textEditor is null)
+            {
+                Debug.WriteLine("The \"textEditor\" argument of \"GoToLine()\" is null.");
```
Comment (Miepee) [inline UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs:420]: ditto

File: UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs [hunk 4]
```diff
@@ -75,6 +74,14 @@ public partial class UndertaleCodeEditor : DataUserControl
         private static readonly Dictionary<string, UndertaleNamedResource> FunctionsDict = new();
         private static readonly Dictionary<string, UndertaleNamedResource> CodeDict = new();
 
+        public enum CodeEditorTab
+        {
+            Unstated,
```
Comment (Miepee) [inline UndertaleModTool/Editors/UndertaleCodeEditor.xaml.cs:79]: i would rename this from "unstated" to "unknown" or "invalid"

**Comment by VladiStep** [comment]
Why I merged "master" into it?

**Review by Miepee — APPROVED** [review]
got no thoughts right now on how to get rid of the tuple, so fine by me with the exception of grammar.
the whole search thing needs to be rewritten in the future anyway. so lets not stall this feature for that.

File: UndertaleModTool/Windows/ClickableTextOutput.xaml.cs [hunk 5]
```diff
@@ -131,7 +135,12 @@ public void GenerateResults()
             headerPara.Inlines.Add(new LineBreak());
             doc.Blocks.Add(headerPara);
 
-            foreach (KeyValuePair<string, List<string>> result in resultsDict)
+            int totalLineCount = resultsDict.Select(x => x.Value.Count).Sum();
+            bool tooManyLines = totalLineCount > 10000;
+            if (tooManyLines)
+                mainWindow.ShowWarning($"There are too many code lines to display ({totalLineCount}), so there would be no clickable line numbers.");
```
[the remark this answers is not in the document]
Comment (VladiStep) [reply UndertaleModTool/Windows/ClickableTextOutput.xaml.cs:141]: Hmm, what if "...so the line numbers won't be clickable."?

[the remark this answers is not in the document]
Comment (Miepee) [reply UndertaleModTool/Windows/ClickableTextOutput.xaml.cs:141]: "won't" (will not) is future. you want present though. 
Whether you use "so the line numbers aren't clickable" or "so no clickable line numbers are shown" doesnt really matter for me.

**Pull request merged**